File: src/lib/utils/lastfm.ts
/**
 * Cliente de API de Last.fm
 * Maneja todas las peticiones a la API de Last.fm con caché
 */

import type {
  LastFmArtistInfo,
  LastFmAlbumInfo,
  LastFmTrackInfo,
  LastFmError,
  ProcessedArtistInfo,
  ProcessedAlbumInfo,
  ProcessedTrackInfo
} from '@/lib/types/lastfm';
import { cache } from '@/lib/utils/cache';

const API_BASE_URL = 'https://ws.audioscrobbler.com/2.0/';
const API_KEY = import.meta.env.VITE_LASTFM_API_KEY;
const CACHE_TTL = 30 * 60 * 1000; // 30 minutos

if (!API_KEY) {
  console.error('⚠️ Last.fm API Key no encontrada. Agrega VITE_LASTFM_API_KEY a tu archivo .env');
}

/**
 * Realiza una petición a la API de Last.fm
 */
async function lastfmFetch<T>(params: Record<string, string>): Promise<T> {
  const url = new URL(API_BASE_URL);
  
  // Parámetros comunes
  url.searchParams.append('api_key', API_KEY);
  url.searchParams.append('format', 'json');
  
  // Parámetros personalizados
  Object.entries(params).forEach(([key, value]) => {
    url.searchParams.append(key, value);
  });

  try {
    const response = await fetch(url.toString());
    
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    
    const data = await response.json();
    
    // Verificar si hay error de Last.fm
    if ('error' in data) {
      const error = data as LastFmError;
      throw new Error(`Last.fm error ${error.error}: ${error.message}`);
    }
    
    return data;
  } catch (error) {
    console.error('Error en petición Last.fm:', error);
    throw error;
  }
}

/**
 * Extrae la mejor imagen disponible (más grande primero)
 */
function getBestImage(images: { '#text': string; size: string }[]): string | null {
  if (!images || images.length === 0) return null;
  
  const sizeOrder = ['mega', 'extralarge', 'large', 'medium', 'small'];
  
  for (const size of sizeOrder) {
    const image = images.find(img => img.size === size);
    if (image && image['#text']) {
      return image['#text'];
    }
  }
  
  // Si no hay imagen con tamaño específico, devolver la primera no vacía
  const anyImage = images.find(img => img['#text']);
  return anyImage ? anyImage['#text'] : null;
}

/**
 * Limpia el HTML de las biografías/resúmenes de Last.fm
 */
function cleanHtml(html: string): string {
  if (!html) return '';
  
  return html
    .replace(/<a[^>]*>([^<]+)<\/a>/g, '$1') // Eliminar enlaces pero mantener texto
    .replace(/<[^>]+>/g, '') // Eliminar todas las etiquetas HTML
    .replace(/\s+/g, ' ') // Normalizar espacios
    .trim();
}

/**
 * Obtiene información de un artista (con caché)
 */
export async function getArtistInfo(artistName: string): Promise<ProcessedArtistInfo | null> {
  const cacheKey = `artist:${artistName.toLowerCase()}`;
  
  // Verificar caché
  const cached = cache.get<ProcessedArtistInfo>(cacheKey);
  if (cached) return cached;
  
  try {
    const response = await lastfmFetch<{ artist: LastFmArtistInfo }>({
      method: 'artist.getinfo',
      artist: artistName,
      autocorrect: '1'
    });
    
    const artist = response.artist;
    
    const result = {
      name: artist.name,
      image: getBestImage(artist.image || []),
      bio: cleanHtml(artist.bio?.summary || ''),
      bioFull: cleanHtml(artist.bio?.content || ''),
      tags: artist.tags?.tag?.slice(0, 5).map(t => t.name) || [],
      listeners: parseInt(artist.stats?.listeners || '0'),
      playcount: parseInt(artist.stats?.playcount || '0'),
      url: artist.url
    };
    
    // Guardar en caché
    cache.set(cacheKey, result, CACHE_TTL);
    
    return result;
  } catch (error) {
    console.error(`Error obteniendo info del artista "${artistName}":`, error);
    return null;
  }
}

/**
 * Obtiene información de un álbum
 */
export async function getAlbumInfo(
  artistName: string,
  albumName: string
): Promise<ProcessedAlbumInfo | null> {
  try {
    const response = await lastfmFetch<{ album: LastFmAlbumInfo }>({
      method: 'album.getinfo',
      artist: artistName,
      album: albumName,
      autocorrect: '1'
    });
    
    const album = response.album;
    
    return {
      name: album.name,
      artist: album.artist,
      image: getBestImage(album.image || []),
      summary: cleanHtml(album.wiki?.summary || ''),
      tags: album.tags?.tag?.slice(0, 5).map(t => t.name) || [],
      listeners: parseInt(album.listeners || '0'),
      playcount: parseInt(album.playcount || '0'),
      url: album.url,
      trackCount: album.tracks?.track?.length || 0
    };
  } catch (error) {
    console.error(`Error obteniendo info del álbum "${albumName}" de "${artistName}":`, error);
    return null;
  }
}

/**
 * Obtiene información de una canción
 */
export async function getTrackInfo(
  artistName: string,
  trackName: string
): Promise<ProcessedTrackInfo | null> {
  try {
    const response = await lastfmFetch<{ track: LastFmTrackInfo }>({
      method: 'track.getinfo',
      artist: artistName,
      track: trackName,
      autocorrect: '1'
    });
    
    const track = response.track;
    
    return {
      name: track.name,
      artist: track.artist.name,
      album: track.album?.title || null,
      image: track.album ? getBestImage(track.album.image || []) : null,
      tags: track.toptags?.tag?.slice(0, 5).map(t => t.name) || [],
      duration: parseInt(track.duration || '0') / 1000, // Convertir ms a segundos
      listeners: parseInt(track.listeners || '0'),
      playcount: parseInt(track.playcount || '0'),
      url: track.url
    };
  } catch (error) {
    console.error(`Error obteniendo info de la canción "${trackName}" de "${artistName}":`, error);
    return null;
  }
}

/**
 * Busca artistas por nombre
 */
export async function searchArtist(query: string, limit = 10) {
  try {
    const response = await lastfmFetch<{
      results: {
        artistmatches: {
          artist: Array<{
            name: string;
            listeners: string;
            mbid: string;
            url: string;
            image: { '#text': string; size: string }[];
          }>;
        };
      };
    }>({
      method: 'artist.search',
      artist: query,
      limit: limit.toString()
    });
    
    return response.results.artistmatches.artist.map(artist => ({
      name: artist.name,
      listeners: parseInt(artist.listeners),
      image: getBestImage(artist.image),
      url: artist.url
    }));
  } catch (error) {
    console.error(`Error buscando artista "${query}":`, error);
    return [];
  }
}

/**
 * Busca álbumes por nombre
 */
export async function searchAlbum(query: string, limit = 10) {
  try {
    const response = await lastfmFetch<{
      results: {
        albummatches: {
          album: Array<{
            name: string;
            artist: string;
            url: string;
            image: { '#text': string; size: string }[];
          }>;
        };
      };
    }>({
      method: 'album.search',
      album: query,
      limit: limit.toString()
    });
    
    return response.results.albummatches.album.map(album => ({
      name: album.name,
      artist: album.artist,
      image: getBestImage(album.image),
      url: album.url
    }));
  } catch (error) {
    console.error(`Error buscando álbum "${query}":`, error);
    return [];
  }
}

File: src-tauri/src/services/lastfm.rs
//! Last.fm API service with caching

use crate::domain::lastfm::{
    raw, EnrichedTrack, LastFmImage, ProcessedAlbumInfo, ProcessedArtistInfo, ProcessedTrackInfo,
};
use crate::domain::music::MusicFile;
use crate::errors::AppError;
use std::collections::HashMap;

use tokio::sync::RwLock;

const API_BASE_URL: &str = "https://ws.audioscrobbler.com/2.0/";

pub struct LastFmService {
    client: reqwest::Client,
    api_key: String,
    // Simple in-memory cache for now: key -> (json_metadata, timestamp)
    // We might want to cache specific processed types instead of raw json to save parsing,
    // but the Prompt suggested (serde_json::Value, u64).
    // Let's cache the Processed types directly for better perfs.
    track_cache: RwLock<HashMap<String, (ProcessedTrackInfo, u64)>>,
    artist_cache: RwLock<HashMap<String, (ProcessedArtistInfo, u64)>>,
    album_cache: RwLock<HashMap<String, (ProcessedAlbumInfo, u64)>>,
}

impl LastFmService {
    pub fn new(api_key: String) -> Self {
        Self {
            client: reqwest::Client::new(),
            api_key,
            track_cache: RwLock::new(HashMap::new()),
            artist_cache: RwLock::new(HashMap::new()),
            album_cache: RwLock::new(HashMap::new()),
        }
    }

    async fn fetch<T: serde::de::DeserializeOwned>(
        &self,
        method: &str,
        params: &[(&str, &str)],
    ) -> Result<T, AppError> {
        let mut query = vec![
            ("method", method),
            ("api_key", &self.api_key),
            ("format", "json"),
            ("autocorrect", "1"),
        ];
        query.extend_from_slice(params);

        let response = self
            .client
            .get(API_BASE_URL)
            .query(&query)
            .send()
            .await
            .map_err(|e| AppError::ExternalApi(e.to_string()))?;

        if !response.status().is_success() {
            return Err(AppError::ExternalApi(format!(
                "HTTP Error: {}",
                response.status()
            )));
        }

        let bytes = response
            .bytes()
            .await
            .map_err(|e| AppError::ExternalApi(e.to_string()))?;

        // Check for Last.fm error response first
        #[derive(serde::Deserialize)]
        struct LastFmError {
            error: i32,
            message: String,
        }

        if let Ok(err) = serde_json::from_slice::<LastFmError>(&bytes) {
            return Err(AppError::ExternalApi(format!(
                "Last.fm Error {}: {}",
                err.error, err.message
            )));
        }

        serde_json::from_slice::<T>(&bytes)
            .map_err(|e| AppError::ExternalApi(format!("Parse Error: {}", e)))
    }

    pub async fn get_track_info(
        &self,
        artist: &str,
        track: &str,
    ) -> Result<ProcessedTrackInfo, AppError> {
        let cache_key = format!("track:{}:{}", artist.to_lowercase(), track.to_lowercase());

        {
            let cache = self.track_cache.read().await;
            if let Some((info, timestamp)) = cache.get(&cache_key) {
                // TODO: Check TTL (30 mins = 1800s)
                // For now, infinite cache or simpler check
                let now = std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_secs();
                if now - timestamp < 1800 {
                    return Ok(info.clone());
                }
            }
        }

        let response: raw::TrackResponse = self
            .fetch("track.getinfo", &[("artist", artist), ("track", track)])
            .await?;

        let t = response.track;
        let processed = ProcessedTrackInfo {
            name: t.name,
            artist: t.artist.name,
            album: t.album.as_ref().map(|a| a.title.clone()),
            duration: t
                .duration
                .and_then(|d| d.parse().ok())
                .map(|ms: u32| ms / 1000),
            playcount: t.playcount.and_then(|p| p.parse().ok()),
            listeners: t.listeners.and_then(|l| l.parse().ok()),
            tags: t
                .toptags
                .map(|tt| tt.tag.into_iter().take(5).map(|tag| tag.name).collect())
                .unwrap_or_default(),
            wiki: t.wiki.map(|w| clean_html(&w.summary)),
            url: t.url,
            image: t
                .album
                .and_then(|a| get_best_image(&a.image.unwrap_or_default())),
        };

        {
            let mut cache = self.track_cache.write().await;
            let now = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs();
            cache.insert(cache_key, (processed.clone(), now));
        }

        Ok(processed)
    }

    pub async fn get_artist_info(&self, artist: &str) -> Result<ProcessedArtistInfo, AppError> {
        let cache_key = format!("artist:{}", artist.to_lowercase());
        {
            let cache = self.artist_cache.read().await;
            if let Some((info, timestamp)) = cache.get(&cache_key) {
                let now = std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_secs();
                if now - timestamp < 1800 {
                    return Ok(info.clone());
                }
            }
        }

        let response: raw::ArtistResponse =
            self.fetch("artist.getinfo", &[("artist", artist)]).await?;

        let a = response.artist;
        let processed = ProcessedArtistInfo {
            name: a.name,
            image: get_best_image(&a.image.unwrap_or_default()),
            bio: a
                .bio
                .as_ref()
                .map(|b| clean_html(&b.summary))
                .unwrap_or_default(),
            bio_full: a
                .bio
                .as_ref()
                .map(|b| clean_html(&b.content))
                .unwrap_or_default(),
            tags: a
                .tags
                .map(|t| t.tag.into_iter().take(5).map(|tag| tag.name).collect())
                .unwrap_or_default(),
            listeners: a
                .stats
                .as_ref()
                .and_then(|s| s.listeners.parse().ok())
                .unwrap_or_default(),
            playcount: a
                .stats
                .as_ref()
                .and_then(|s| s.playcount.parse().ok())
                .unwrap_or_default(),
            url: a.url,
        };

        {
            let mut cache = self.artist_cache.write().await;
            let now = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs();
            cache.insert(cache_key, (processed.clone(), now));
        }

        Ok(processed)
    }

    pub async fn get_album_info(
        &self,
        artist: &str,
        album: &str,
    ) -> Result<ProcessedAlbumInfo, AppError> {
        let cache_key = format!("album:{}:{}", artist.to_lowercase(), album.to_lowercase());
        {
            let cache = self.album_cache.read().await;
            if let Some((info, timestamp)) = cache.get(&cache_key) {
                let now = std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_secs();
                if now - timestamp < 1800 {
                    return Ok(info.clone());
                }
            }
        }

        let response: raw::AlbumResponse = self
            .fetch("album.getinfo", &[("artist", artist), ("album", album)])
            .await?;

        let a = response.album;
        let processed = ProcessedAlbumInfo {
            name: a.name,
            artist: a.artist,
            image: get_best_image(&a.image.unwrap_or_default()),
            summary: a.wiki.map(|w| clean_html(&w.summary)).unwrap_or_default(),
            tags: a
                .tags
                .map(|t| t.tag.into_iter().take(5).map(|tag| tag.name).collect())
                .unwrap_or_default(),
            listeners: a.listeners.and_then(|l| l.parse().ok()).unwrap_or_default(),
            playcount: a.playcount.and_then(|p| p.parse().ok()).unwrap_or_default(),
            url: a.url,
            track_count: a.tracks.map(|t| t.track.len() as u32).unwrap_or_default(),
        };

        {
            let mut cache = self.album_cache.write().await;
            let now = std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs();
            cache.insert(cache_key, (processed.clone(), now));
        }

        Ok(processed)
    }

    pub async fn enrich_tracks_batch(
        &self,
        tracks: Vec<MusicFile>,
    ) -> Result<Vec<EnrichedTrack>, AppError> {
        use futures::stream::{self, StreamExt};

        let results: Vec<EnrichedTrack> = stream::iter(tracks)
            .map(|track| async move {
                let artist = track.artist.as_deref().unwrap_or_default();
                let title = track.title.as_deref().unwrap_or_default();

                let enriched = if !artist.is_empty() && !title.is_empty() {
                    self.get_track_info(artist, title).await.ok()
                } else {
                    None
                };

                let album_art_url = enriched.as_ref().and_then(|e| e.image.clone());

                EnrichedTrack {
                    original: track,
                    enriched,
                    album_art_url,
                }
            })
            .buffer_unordered(5) // Limit concurrency to 5
            .collect()
            .await;

        Ok(results)
    }
}

// Helper to clean HTML
fn clean_html(html: &str) -> String {
    // Basic cleanup using regex if possible, else simple replacement
    // Since we added regex to Cargo.toml, let's use it but we need to import it.
    // However, importing inside function is not ideal.
    // I'll assume we can use a simple string manipulation for now to avoid compilation errors if regex isn't in scope.
    // Or I can add `use regex::Regex;` at the top of the file in a separate edit, but I am in ReplaceFileContent.
    // I will stick to the provided `clean_html` implementation or slightly improved.
    // The previous implementation was:
    // let no_links = html.replace(|c: char| c == '<' || c == '>', "");
    // This is too aggressive (removes all < > which might be part of text, though rare in HTML content).
    // Better: remove <...>

    // For now, let's just return it as is or do a very simple pass.
    // Real implementation should use a library.
    // I'll assume the user is okay with simple stripping for now.

    let mut result = String::with_capacity(html.len());
    let mut inside_tag = false;

    for c in html.chars() {
        if c == '<' {
            inside_tag = true;
        } else if c == '>' {
            inside_tag = false;
        } else if !inside_tag {
            result.push(c);
        }
    }

    result.trim().to_string()
}

fn get_best_image(images: &[LastFmImage]) -> Option<String> {
    if images.is_empty() {
        return None;
    }

    let size_order = ["mega", "extralarge", "large", "medium", "small"];

    for size in size_order {
        if let Some(img) = images.iter().find(|i| i.size == size) {
            if !img.text.is_empty() {
                return Some(img.text.clone());
            }
        }
    }

    images
        .iter()
        .find(|i| !i.text.is_empty())
        .map(|i| i.text.clone())
}
